### ISSUE TITLE
interest in modernizing?

### ISSUE THREAD

**Issue by jGaboardi** [carsonfarmer/fastpair#20]
@carsonfarmer Do you have any interest in seeing `fastpairs` modernized (e.g., Python 312 compatibility, `pyproject.toml`, GHA for testing, structured formatting & linting, etc.)? If so (and as time permits), I can start making specific tickets and chipping away at it.

xref: #12

**Comment by carsonfarmer** [comment]
I mean, yeh awesome. That would be amazing!

**Comment by jGaboardi** [comment]
### Plan of attack

1. initialize a single `py312_latest` bare-bones environment + GHA for testing current package / code base - ensure passing
   1. `.github/workflows/testing.yml`
   2. `ci/py312_latest.yml`
2. remove `.travis.yml`
3. consolidate old infrastructure into `pyproject.toml`
   1. `setup.py`
   2. `requirements.txt`
   3. `recommended.txt`
   4. `.coveragerc`
5. (turn on pre-commit & codecov for the repo)?
4. add `ruff` specifications to `pyproject.toml` for linting and formatting codebase
6. remove Python 2 adaptations in the code base
7. add a `.pre-commit-configuration.yaml`
8. fill out CI envs
   1. `py310_oldest` (based on [SPEC000](https://scientific-python.org/specs/spec-0000/))
   2. `py311_latest`
   3. `py312_dev`
9. fill out testing with any edge cases that crop up
10. add a `notebooks/` or `examples/` directory with some demonstrations
11. revisit docstrings for [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) conformity
12. revisit `README` for updates, badges, etc.
13. (potential) see about `docs/` + a GHPages site built from a GHA
14. (potential) see about release GHA based on pushed `v*` tags (automated publishing on PyPI)
15. Following release on PyPI, initialize release on `conda-forge` as a feedstock


@carsonfarmer This is a loosely sequential checklist of proposed activities. Some make sense to do in individual PRs and some may be grouped. Let me know if you have any questions or concerns with anything here, otherwise I'll see about getting started with (1.)

**Comment by carsonfarmer** [comment]
TL;DR: This all looks good to me!

I'm pretty unfamiliar with modern Python development flows at this stage @jGaboardi, so I would defer to your judgement here. Since I have, in fact, met you in person, I have done ahead and invited you to be a collaborator on this repo. You should have direct access to GitHub Actions etc as needed with that invite?

**Comment by carsonfarmer** [comment]
Totally separate note: Is PySAL _using_ this library?

**Comment by jGaboardi** [comment]
> Totally separate note: Is PySAL _using_ this library?

Not currently, but I became aware of `fastpair` last summer when @ljwolf, @gegen07, and I were working with a student in [a PySAL GSoC project for `spopt`](https://summerofcode.withgoogle.com/archive/2023/projects/05jDO3IG). So there's potential for inclusion later down the line in our new [`KNearestPMedian`](https://pysal.org/spopt/generated/spopt.locate.KNearestPMedian.html#spopt.locate.KNearestPMedian) model, but we'll need to do performance testing, benchmarking, etc. before that happens. So, my contributions here are for 3 reasons (in order of general to specific):
1. I love the packaging aspect of the Scientific Python ecosystem.
2. I believe your work on this has a lot value and making it available to a broader audience will hopefully make others' lives easier.
3. The specific `pysal/spopt` reason mentioned above.

**Comment by gegen07** [comment]
@jGaboardi I saw a lot of work done by you! Awesome! Any help wanted? 
The spopt will be helped with this brilliant implementation, thanks @carsonfarmer!

**Comment by jGaboardi** [comment]
>  Any help wanted?

Of course! Your insight is always valuable, specifically due to your CS background. Anything specific that has caught your eye in the issues?

I can't speak for @carsonfarmer but IMHO we should get `fastpair` modernized with an actual release before addressing enhancements like:
* #34 
* #39 
* #42

**Comment by gegen07** [comment]
No problem with current issues. All good!
I'll take a look at the issues for in-depth enhancements.
Agree with you, before enhancing it, let's get it over the release cycle first.
